### ISSUE TITLE
Keycloak login failure when replicas > 1

### ISSUE THREAD

**Issue by oridool** [codecentric/helm-charts#149]
When I set keycloak.replicas > 1, I cannot login to web console. 
After clicking the username/password, it redirects again to the login screen.
In logs, there is the following error:

16:34:16,486 WARN  [org.keycloak.events] (default task-2) type=LOGIN_ERROR, realmId=master, clientId=null, userId=null, ipAddress=194.90.225.116, error=invalid_code

I managed to **solve** the issue by enable cookie stickiness in the ngnix and set in Values.yaml:
ingress.annotations.nginx.ingress.kubernetes.io/affinity: "cookie"

1) I suggest that in the readme.md file, you'll add this note. Whenever replicas is set to value greater than 1, you need also to enable cookie affinity in ingress.
2) not sure it is an issue, but the original chart key was "annotations.nginx.ingress.kubernetes.io/affinity", while I had to add the "ngnix" prefix: "ngnix.annotations.nginx.ingress.kubernetes.io/affinity.

**Comment by campenberger** [comment]
I have the problem. replicas is set to 2, but the jgroup variables are not configured in the generated statefulset:
```
# Source: keycloak/templates/statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: keycloak-cc
  labels:
    app.kubernetes.io/name: keycloak
    helm.sh/chart: keycloak-6.0.3
    app.kubernetes.io/instance: keycloak-cc
    app.kubernetes.io/managed-by: Tiller
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: keycloak
      app.kubernetes.io/instance: keycloak-cc
  replicas: 2
  serviceName: keycloak-cc-headless
  podManagementPolicy: Parallel
  updateStrategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        app.kubernetes.io/name: keycloak
        app.kubernetes.io/instance: keycloak-cc
      annotations:
        checksum/config-sh: c79822799f71bdd2ded30ba4f74a0eaf5a900df6652ea919c9defda8524dc997
        checksum/config-startup: 77507a988c74ee0186ccdd21bb7bc28999a2854e571891655d68435214912b70
    spec:
      restartPolicy: Always
      serviceAccountName: default
      securityContext:
        fsGroup: 1000
        
      containers:
        - name: keycloak
          image: "jboss/keycloak:7.0.1"
          imagePullPolicy: IfNotPresent
          securityContext:
            runAsNonRoot: true
            runAsUser: 1000
            
          command:
            - /scripts/keycloak.sh
          lifecycle:
            # postStart:
            #   exec:
            #     command: ["/bin/sh", "-c", "ls"]
            
          env:
            - name: KEYCLOAK_USER
              value: admin
            - name: KEYCLOAK_PASSWORD_FILE
              value: /secrets/password
            
            - name: DB_VENDOR
              value: "postgres"
            - name: DB_ADDR
              value: "pgsql"
            - name: DB_PORT
              value: "5432"
            - name: DB_DATABASE
              value: "keycloak"
            - name: DB_USER
              value: "postgres"
            - name: DB_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: keycloak-pgsql
                  key: "password"
            - name: PROXY_ADDRESS_FORWARDING
              value: "true"
            # - name: KEYCLOAK_LOGLEVEL
            #   value: DEBUG
            # - name: WILDFLY_LOGLEVEL
            #   value: DEBUG
            # - name: CACHE_OWNERS
            #   value: "2"
```
The JGROUPS_ variables should be right after the KEYCLOAD_PASSWORD_FILE

**Comment by dojadop** [comment]
@oridool @campenberger what version of helm are you using? there was a bug with helm 2.15.0. 

See my comment here: https://github.com/codecentric/helm-charts/issues/130

**Comment by unguiculus** [comment]
Thanks @dojadop for jumping in. That could very well be there reason for the issue.

**Comment by oridool** [comment]
@dojadop , thanks for that note. I'm using helm 2.11, probably this is why I had the problem.
As I mentioned, the workaround (if not upgrading helm) is modifying "annotations.nginx.ingress.kubernetes.io/affinity".

**Comment by unguiculus** [comment]
I guess we have two different issues here. Yours seems different from @campenberger's.

**Comment by campenberger** [comment]
you were correct, my issue is because of the helm version. It works now. Sorry for not finding #130 and thank you for your help

**Comment by unguiculus** [comment]
@dojadop This should not be related to the Helm version. After all, Helm just renders and applies manifests. Could you check the logs if JGroups discovery works?

**Comment by dojadop** [comment]
@unguiculus there was a regression in helm 2.15.0 (https://github.com/helm/helm/issues/6708) where it didn't treat numeric value types appropriately cause the JGROUPS env variables to not be rendered in the output

**Comment by bbfeng1979** [comment]
我采用的 Rancher 来管理集群，如果直接增加 keycloak 的个数，就会出现这种问题，登陆 keycloak 报 ERR_TOO_MANY_REDIRECTS。

这是由于 keycloak 在启动的时候，会去加载编排 scripts 下面的几个文件，比如 ha.cli 这种，如果直接增加个数，就绕开了这些配置。

因此，如果要增加个数，通过升级的方式，先设置好 replicas 个数在升级，就没有这个问题了。

**Comment by brunocascio** [comment]
Hi Guys,

According to https://www.keycloak.org/2019/05/keycloak-cluster-setup for K8s you have to use KUBE_PING Discovery Protocol if you want to run keycloak in cluster mode (replicas > 1).

Also, keycloak recommends set sticky option in your the lb/proxy in order to improve the cache performance but it's not the solution for replicas of course.

Cheers!


TL;DR;

Keycloak is working in a new architecture https://www.keycloak.org/2019/10/keycloak-x

**Comment by Jaorji** [comment]
I also have a problem here, when I have replica=2, sometimes I will get 403 get access to resources. I'm trying to see if @oridool 's solution works for me. However, I'm using istio in my project, do you know how to enable cookie affinity in ingress in istio?

**Comment by nrx-ops** [comment]
I experienced same problem.

**Comment by nrx-ops** [comment]
I experienced same issue.
kubernetes.KUBE_PING works well.
```
extraEnv: |
  - name: JGROUPS_DISCOVERY_PROTOCOL
    value: kubernetes.KUBE_PING
  - name: JGROUPS_DISCOVERY_PROPERTIES
    value: dump_requests=true,port_range=0,namespace=<your_KC_namespace>
```
If JGROUPS_DISCOVERY_PROPERTIES is not properly set, each pod consider him as master and generate an SQL Error and/or ERR_TOO_MANY_REDIRECTS.

This config works without sticky session workaround.
Thank to @brunocascio

**Comment by unguiculus** [comment]
Closing this as resolved now.
